Sources/GRPC/ConnectionPool/PoolManager.swift

```diff
@@ -0,0 +1,270 @@
+  ///       The `EventLoop` of the selected pool will be the same as the `EventLoop` of
+  ///       the `EventLoopFuture<Channel>` returned from this call.
+  ///   - deadline: The point in time by which the stream must have been selected. If this deadline
+  ///       is passed then the returned `EventLoopFuture` will be failed.
+  ///   - logger: A logger.
+  ///   - initializer: A closure to initialize the `Channel` with.
+  /// - Returns: An `EventLoopFuture<Channel>` on the same `EventLoop` as the connection pool
+  ///     providing the `Channel`. The future will be failed if the pool manager has been shutdown,
+  ///     the deadline has passed before a stream was created or if the selected connection pool
+  ///     is unable to create a stream (if there is too much demand on that pool, for example).
+  internal func makeStream(
+    preferredEventLoop: EventLoop?,
+    deadline: NIODeadline,
+    logger: GRPCLogger,
+    streamInitializer initializer: @escaping (Channel) -> EventLoopFuture<Void>
+  ) -> EventLoopFuture<Channel> {
+    let reservation = self.lock.withLock {
+      self.state.reserveStream(preferringPoolOnEventLoop: preferredEventLoop)
+    }
+
+    switch reservation {
+    case let .success(pool):
+      return pool.makeStream(deadline: deadline, logger: logger, initializer: initializer)
+
+    case let .failure(error):
+      return (preferredEventLoop ?? self.group.next()).makeFailedFuture(error)
+    }
+  }
+
+  // MARK: Shutdown
+
+  /// Shutdown the pool manager and all connection pools it manages.
+  internal func shutdown() -> EventLoopFuture<Void> {
+    let promise = self.group.next().makePromise(of: Void.self)
+    self.shutdown(promise: promise)
+    return promise.futureResult
+  }
+
+  /// Shutdown the pool manager and all connection pools it manages.
+  internal func shutdown(promise: EventLoopPromise<Void>) {
+    let action = self.lock.withLock {
+      return self.state.shutdown(promise: promise)
+    }
+
+    switch action {
+    case let .shutdownPools(pools):
+      promise.futureResult.whenComplete { _ in self.shutdownComplete() }
+      EventLoopFuture.andAllSucceed(pools.map { $0.shutdown() }, promise: promise)
+
+    case let .alreadyShuttingDown(future):
+      promise.completeWith(future)
+
+    case .alreadyShutdown:
+      promise.succeed(())
+    }
+  }
+
+  private func shutdownComplete() {
+    self.lock.withLockVoid {
+      self.state.shutdownComplete()
+    }
+  }
+}
+
+// MARK: - Connection Pool to Pool Manager
+
+extension PoolManager: StreamLender {
+  internal func returnStreams(_ count: Int, to pool: ConnectionPool) {
+    self.lock.withLockVoid {
+      self.state.returnStreams(count, to: pool)
+    }
+  }
+
+  internal func increaseStreamCapacity(by delta: Int, for pool: ConnectionPool) {
+    self.lock.withLockVoid {
+      self.state.increaseStreamCapacity(by: delta, for: pool)
+    }
+  }
```

Comment:
Could a streams capacity also be decreased?

Reply:
Yes, a negative `delta` is acceptable here. I'll change this to `changeStreamCapacity`.